Fixes #18426 - Add actions to notification entries

dLobatog opened the pull request:
This adds support for the notifications to display actions. In case no
actions are provided, the dropdown kebab isn't shown. The actions
possible right now only include GET links.

![screenshot from 2017-02-09 12-16-25](https://cloud.githubusercontent.com/assets/598891/22781213/c85d6b6c-eec1-11e6-9c2e-d7e435146103.png)
![screenshot from 2017-02-09 12-16-44](https://cloud.githubusercontent.com/assets/598891/22781214/c892dfc2-eec1-11e6-95e0-6435436d5646.png)

dLobatog committed:
Fixes #18426 - Add actions to notification entries

This adds support for the notifications to display actions. In case no
actions are provided, the dropdown kebab isn't shown. The actions
possible right now only include GET links.

ohadlevy commented:
@gailsteiger or @matanwerbner can you have a look please?

ohadlevy commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+    }
+
+    actions_dropdown = (
+      <div className="dropdown pull-right dropdown-kebab-pf">
```
does it make sense to make kebab a component by it self? i can see us using it in many different places

gailsteiger replied:
yes

ohadlevy commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+
+    actions_dropdown = (
+      <div className="dropdown pull-right dropdown-kebab-pf">
+        <button className="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight{id}"
```
maybe it make sense to use https://react-bootstrap.github.io/components.html#btn-dropdowns ?

matanwerbner replied:
i think we can implement this component ourself using a stateful React class, more reason for it to be on it's own

gailsteiger replied:
We should implement the [kebab](http://www.patternfly.org/pattern-library/widgets/#kebabs) as a stand alone component using the patternfly markup as a reference. We probably should remove the bootstrap js functionality and replace with a pure react implementation.

matanwerbner replied:
thinking of it, the kebab is just an icon, so perhaps we can have it as a css class icon, with onClick to open various panels as component requires

gailsteiger replied:
In patternfly, the kebab is a restyled version of a dropdown supporting the same features as the bootstrap dropdown component. I'll check with Roxanne as to whether there are use cases for the dropdown menu or whether the kebab should always be used.

matanwerbner commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+  const links = [];
+
+  let actions_dropdown = '';
+
```
This component may best be put on it's own as a seperate component, accepting a list of links, and then on the parent's render:
{ actions.links && `<ActionsDropdown  links={actions.links } />`   }
that way we dont inflate the notification component too much, and dont have to deal with different types of notifications

gailsteiger replied:
yes

matanwerbner commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+  let actions_dropdown = '';
+
+  if (actions.links != null) {
+    for (let i = 0; i < actions.links.length; i++) {
```
`const links = actions.links.map((link, index) => {
  const li_ley = `notifcation-link${index}-${id}`;
  return <li key={li_key} id={ li_ley }>
                 <a target="_blank" href={link.href}>
                      {link.title}
                 </a></li>)`

matanwerbner replied:
i think we can easily implement this component ourself using a stateful component, more reason for it to be on it's own

matanwerbner commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+  let actions_dropdown = '';
+
+  if (actions.links != null) {
+    for (let i = 0; i < actions.links.length; i++) {
```
it might be nicer to use array.map 
`const links = actions.links.map((link, index) => {

})`

gailsteiger replied:
+1 for using `map`

matanwerbner commented on webpack/assets/javascripts/react_app/components/notifications/Notification.js:
```diff
@@ -18,6 +18,33 @@ const Notification = ({created_at, seen, text, level, id}) => {
+  const links = [];
+
+  let actions_dropdown = '';
+
```
agree. also, shouldn't there be a onClick handler here?

dLobatog commented:
@ohadlevy @gailsteiger @matanwerbner 

Updated - the Dropdown is now in its own component. It accepts the ID in order to create unique ids for the elements (which QE deems useful). I could see making it more general but I think it's best to leave it 'YAGNI' and add it when we actually want it.

Aside from that, I believe I addressed the other comments (using map instead of the for iterator, etc..)

matanwerbner commented on webpack/assets/javascripts/react_app/components/common/Dropdown.js:
```diff
@@ -0,0 +1,29 @@
+
+const Dropdown = ({ links, id }) => {
+  const listLinks = [];
+
```
functional approach to writing this would be:

```

const listLinks = links.map((link, i) => { 
  const liKey = `notification-link-${i}-${id}`
 return <li key={liKey} id={liKey}>
                    <a target="_blank" href={link.href}>
                    {link.title}</a>
                  </li>
}
```

dLobatog replied:
That's true - haste is the enemy of doing a good job 😄  Will update in a moment

ohadlevy commented on app/models/notification_recipient.rb:
```diff
@@ -17,6 +17,7 @@ def payload
       :subject    => notification_blueprint.subject,
       :created_at => notification.created_at,
       :group      => notification_blueprint.group,
+      :actions    => notification_blueprint.actions
```
how does the actions translate to actual urls given the subject is part of the notification vs notification_blueprint object?

dLobatog replied:
Actions have nothing to do with subject - it's just a dumb set of links for the moment, it can be any hash https://github.com/theforeman/foreman/pull/4265/files#diff-61072566998937ff274a277f94d284e8R12

It can be made more complicated but I'm not sure how good is it to start coding logic nothing is using, this approach works well with the RSS PR (if passing an array of links isn't good enough ,e.g: you want to pass the HTTP action - POST, etc.., that's easily changeable if we need to do it)

gailsteiger commented on webpack/assets/javascripts/react_app/components/common/Dropdown.js:
```diff
@@ -0,0 +1,27 @@
+      <button className="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight{id}"
+               data-toggle="dropdown" aria-haspopup="true" aria-expanded="true">
+        <span className="fa fa-ellipsis-v"></span>
+      </button>
```
The component is still using bootstrap's dropdown.js plugin to handle the toggle. This is not desirable

dLobatog replied:
Why not? I see it's handled like that on Patternfly and the Angular implementation of Patternfly too.

gailsteiger replied:
Patternfly itself doesn't implement a javascript framework so it makes sense that it uses the bootstrap jquery plugins.

With Angular and React the paradigm moves away from direct manipulation markup and adopts the approach of rendering to reflect state. With React the concept of one way data flow is also introduced.

If you have a look at [AngularUI Bootstrap](https://angular-ui.github.io/bootstrap/) (the port of Bootstrap to Angular) you can see that it contains native AngularJS directives and has no dependency on either jQuery or Bootstrap's JS. I don't know why Angular Patternfly chose to not to implement in the same manner.

I think that our components will be easier to reason about if we adopt the pure React approach instead of mixing Reach with the Bootstrap plugins. Specifically, the bootstrap approach was not workable with the Navigation Drawer and Alert/Toasts (which you can see when I update the PR).

I think would be better to encapsulate a click handler in the component rather than rely on the boostrap plugins. In that way it will be easier for us to manage state, render in the React way and initiate actions.

@matanwerbner what do you think?

matanwerbner replied:
Agree. We should try to not depend on bootstrap at all, especially on simple or trivial components. A bit of css wil get us where we need.

dLobatog commented:
@matanwerbner @gailsteiger Pushed an updated version that uses the `Dropdown` component from react-bootstrap, it's quite useful. 

I'll submit this implementation to https://idimaster.github.io/patternfly-react/ see if we can hopefully rely on this npm package for every Patternfly component 😄

matanwerbner commented on webpack/assets/javascripts/react_app/components/notifications/NotificationDropdown.js:
```diff
@@ -0,0 +1,28 @@
+  });
+
+  return (
+    <Dropdown className="pull-right dropdown-kebab-pf" pullRight key={id}
```
small comment on style, styling with flex box is easier then pull-right/pull-left... if the containing div has display:flex and justify-content:space-between, the elements drop into position nicely :)

dLobatog replied:
I seem to not be able to do this properly as flex-box centers the text 😞

matanwerbner replied:
It's ok, it was just something i thought is the responsibility of the composer (wrapper) component to organize it's children - and not for the child to be opinionated about it's location.

i suggest reading up on flexbox anyway, you may find it easier to deal with then floats

gailsteiger replied:
@dLobatog you might find this useful [A Complete Guide to Flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)

gailsteiger commented:
@dLobatog I don't like react-bootstrap. I think it's overly complicated and tedious to use. I'm not going to object to the PR but I think a simple component would be better.

In this case, you had to use many element tags with the dropdown:
```js
    <Dropdown className="pull-right dropdown-kebab-pf" pullRight key={id}
              id={`notifications-dropdown-${id}`}>
      <Dropdown.Toggle noCaret bsStyle="link">
        <Glyphicon bsClass="fa" glyph="ellipsis-v" />
      </Dropdown.Toggle>
      <Dropdown.Menu>
        {listLinks}
      </Dropdown.Menu>
    </Dropdown>
```
I would have to consult the documentation every time I used it.

Instead, I think we should be able to achieve something along the lines of 
```js
    <Dropdown type="kebab" key={id}  id={`notifications-dropdown-${id}`}>
        {listLinks}
    </Dropdown>
```

PS: I'm not opposed to using component libraries but some are better than others

dLobatog commented:
@gailsteiger thanks for taking a look! About react-bootstrap, I thought it would be a good idea to use it since it is already a dependency of Foreman and the component is fully implemented there. This means different types of toggle, dividers, etc.. we don't have to write them. Additionally the docs are pretty good and with a large community I would expect it to not have many bugs or at least have them fixed at a faster rate than if I wrote our own foreman notifications dropdown component (we can always contribute back 😄 ). 

@matanwerbner Is the flex-box CSS thing a deal-breaker? I'm no expert or anything but it's what I see the Patternfly implementors use in their examples (http://www.patternfly.org/pattern-library/communication/notification-drawer/#/_code) so I suppose they code Patternfly with the expectations pull-direction is used instead of flex-box.

@gailsteiger @matanwerbner ++ appreciate the feedback!

edit - react-bootstrap, not react-patternfly

gailsteiger commented:
@dLobatog I presume your above comments are about react-bootstrap and not about react-patternfly.
I am guilty of having introduced the react-bootstrap dependency into foreman. At present I think we use only one component - modal. I tried a few of the other components and removed them for various reasons.
For the purpose of this PR, leave the react-bootstrap Dropdown in.

I still would like to see the feature running. Could you refer me to a live site or help me get the appropriate data into my database.

dLobatog commented:
@gailsteiger If you want to test this, you can create notifications this way:

1st - Run `rake db:migrate` to run the migration from this pull request that adds 'actions' to the Notifications table.
2nd - Open the Rails console with `rake console` from the Foreman directory
3rd - Issue this commands:
```ruby
blueprint = NotificationBlueprint.new(:group => 'React devs',
                                      :message => 'Hi Gail! This is a notification message',
                                      :level => 'info',
                                      :expires_in => 10.months,
                                      :name => 'gailnotification',
                                      :actions => {"links"=>[{"href"=>"https://theforeman.org/blog", "title"=>"Link to blog"}]})

blueprint.save
notification = Notification.new(:initiator => User.first, :audience => Notification::AUDIENCE_GLOBAL, :notification_blueprint => blueprint)
notification.save
```

It should automatically show up on the UI after that.

matanwerbner commented:
@dLobatog seems good from my end, 
I feel like the structure of the notification drawer can be greatly improved, but that is for another PR

gailsteiger commented:
@dLobatog thanks for the rails console instructions. Looks fine by me.

ohadlevy commented on webpack/assets/javascripts/react_app/components/notifications/NotificationDropdown.js:
```diff
@@ -0,0 +1,28 @@
+    <Dropdown className="pull-right dropdown-kebab-pf" pullRight key={id}
+              id={`notifications-dropdown-${id}`}>
+      <Dropdown.Toggle noCaret bsStyle="link">
+        <Glyphicon bsClass="fa" glyph="ellipsis-v" />
```
why not use our Icon component?

dLobatog replied:
I did not see any advantage on using that component vs using what's provided by react-bootstrap, is there any?

ohadlevy commented:
@matanwerbner can to explain regarding notification structure comment?

ohadlevy commented:
@dLobatog note that a test has failed, it does look like its un-related, but if you push again it would be nice to confirm.

dLobatog commented:
Repushed to confirm the error some node tests in Travis is not this PR's fault

ohadlevy commented:
thanks @dLobatog, katello tests failures are unrelated to this pr.
